### ISSUE TITLE
Native data and metadata encryption for ZFS

### ISSUE THREAD

**Issue by marcheschi** [erigones/esdc-ce#316]
Hi
This is a request and not an issue,
Recent EU directive on data protection enforces "protection measures should include those that render the data unintelligible to any person who is not authorized to access it, in particular by encrypting the personal data"
http://data.consilium.europa.eu/doc/document/ST-9565-2015-INIT/en/pdf
This is enforced specially in europe and in italy.
Zfs, recently added this feature:
https://github.com/openzfs/openzfs/pull/489

I think could be a very big feature for Danubecloud.
Thank you
Paolo

**Comment by YanChii** [comment]
Hi @marcheschi, 
Thank you for the suggestion. We'll take look at it.
But anyway, we have to wait until this openzfs commit gets integrated into SmartOS.
Jan

**Comment by dn0** [comment]
I agree, this would be awesome. Hope it gets integrated into SmartOS.

**Comment by YanChii** [comment]
See more info on progress of crypto integration in recent mailing list messages:
https://www.listbox.com/member/archive/184463/2018/04/sort/time_rev/page/1/entry/2:34/20180416124131:016E9238-4195-11E8-ABCB-BF55BB25E7F7/

**Comment by marcheschi** [comment]
Hi
Finally a stable version of ZFS encryption has been done on illumos, and it is possible to try.
You can download the patch from here:
https://patch-diff.githubusercontent.com/raw/openzfs/openzfs/pull/489.patch

in order to apply it : 

cd smartos-live/projects/illumos
wget https://patch-diff.githubusercontent.com/raw/openzfs/openzfs/pull/489.patch
patch -p1 -i 489.patch

I'm trying it now
Paolo

**Comment by YanChii** [comment]
Hi @marcheschi,
As you have been told in `smartos-discuss`, it's not yet safe nor fully implemented. I am eager to have it, too. But we have to wait until it lands usably in SmartOS.
Anyway, if you care to share the experience from your testing, we will be happy to look at it.
Cheers.
Jan

**Comment by marcheschi** [comment]
Hi @YanChii 
in Smartos-discuss they told me :

> "it's not entirely safe on illumos, and with SmartOS in particular creates some additional challenges"

But it is a very generic assertion, they have to solve still some bugs, If you read the document they [shared](https://github.com/joyent/rfd/blob/master/rfd/0077/README.adoc) with me It is a very interesting document on various kind of attacks and countermeasure, and the proof they are using and investing a lot in this feature.

I aggree with you that it is not production ready, but ready enough for testing.
I will let you know further progress, and try to share my experience.
Best Regards
Paolo

**Comment by marcheschi** [comment]
Hi 
We finally have the zfs encryption feature ready:
https://www.illumos.org/issues/8727#change-25651
Best Regards
Paolo

**Comment by YanChii** [comment]
Cool.
So the minimum requirement has been met. Now we have to find out how to implement it on a higher level. AFAIK there's no `vmadm` integration yet and I don't even see any work in that direction (besides the [RFD](https://github.com/joyent/rfd/blob/master/rfd/0077/README.adoc) but it doesn't help that much). 

Maybe we will need to create our own implementation of key handling.

Jan

**Comment by marcheschi** [comment]
Hi @YanChii 
I need to create a linux VM with encrypted filesystem,
What is the best way to continue to use DC for Administration, backups,..
Thank you
Paolo
